src/plugins/graph/adornments/movable-line/movable-line.tsx

```diff
@@ -325,90 +372,114 @@ export const MovableLine = observer(function MovableLine(props: IProps) {
-        .on("end", endTranslate),
-      lower: drag()
-        .on("start", (e) => startRotation(e, "lower"))
-        .on("drag", (e) => continueRotation(e, "lower"))
-        .on("end", (e) => endRotation()),
-      upper: drag()
-        .on("start", (e) => startRotation(e, "upper"))
-        .on("drag", (e) => continueRotation(e, "upper"))
-        .on("end", (e) => endRotation()),
-      equation: drag()
-        .on("drag", moveEquation)
-        .on("end", endMoveEquation)
-    };
-
-    lineObject.cover?.call(behaviors.cover);
-    lineObject.handleLower?.call(behaviors.lower);
-    lineObject.handleUpper?.call(behaviors.upper);
-    lineObject.equation?.call(behaviors.equation);
-  }, [lineObject, readOnly, continueTranslate, endTranslate,
-    startRotation, continueRotation, endRotation, moveEquation, endMoveEquation]);
-
-  // Build the line and its cover segments and handles just once
+    }, [plotWidth, plotHeight, positionEquation, model]),
+
+    endMoveEquation = useCallback((index: number) => {
+      model.saveEquationCoords(index);
+    }, [model]);
+
+  const addBehaviors = useCallback(() => {
+    lineObjects.current.forEach((lineObject, index) => {
+      const behaviors: {
+        cover: DragBehavior<SVGLineElement, unknown, unknown>,
+        lower: DragBehavior<SVGCircleElement, unknown, unknown>,
+        upper: DragBehavior<SVGCircleElement, unknown, unknown>,
+        equation: DragBehavior<HTMLDivElement, unknown, unknown>
+      } = {
+        cover: drag<SVGLineElement, unknown>()
+          .on("drag", (e) => continueTranslate(e, index))
+          .on("end", () => endTranslate(index)),
+        lower: drag<SVGCircleElement, unknown>()
+          .on("start", (e) => startRotation(e, "lower", index))
+          .on("drag", (e) => continueRotation(e, "lower", index))
+          .on("end", () => endRotation(index)),
+        upper: drag<SVGCircleElement, unknown>()
+          .on("start", (e) => startRotation(e, "upper", index))
+          .on("drag", (e) => continueRotation(e, "upper", index))
+          .on("end", () => endRotation(index)),
+        equation: drag<HTMLDivElement, unknown>()
+          .on("drag", (e) => moveEquation(e, index))
+          .on("end", () => endMoveEquation(index))
+      };
+
+      lineObject.cover?.call(behaviors.cover);
+      lineObject.cover?.on("click", () => toggleLineSelection(index));
+      lineObject.cover?.on("mouseover", () => updateClasses(lineObject.line!, index, true));
+      lineObject.cover?.on("mouseout", () => updateClasses(lineObject.line!, index, false));
+      lineObject.handleLower?.call(behaviors.lower);
+      lineObject.handleUpper?.call(behaviors.upper);
+      lineObject.equation?.call(behaviors.equation);
+    });
+  }, [continueRotation, continueTranslate, endMoveEquation, endRotation, endTranslate, moveEquation, startRotation,
+      toggleLineSelection, updateClasses]);
+
+  // Build the lines and their cover segments and handles just once
```

Comment:
Does this build the d3 objects just once?  It looks like the useEffect / autorun may continually delete and re-add them. I didn't trace the logic completely though.

Reply:
You're right, this can and probably will often run more than once now. I believe that's necessary. I'll remove "just once" from the comment.